ENH: Kstest exact performance improvements

Balthasar-eu opened the pull request:
#### Reference issue
No Issue

#### What does this implement/fix?
This adds two benchmarks, one for scipy.stats.kstest and one for scipy.stats.ks_2samp, and it increases performance of the exact KS test, when an unequal amount of samples are in each group. On my own system it reduces the runtime of the exact KS test benchmars for 20 and 21 samples from ~800µs to ~200µs, while still passing all tests.

I also exchanged the manual test for infinity for a try except statement to catch runtime warnings.

#### Additional information
This is my first pull request here, so I am not sure if everything is done correctly, please let me know if anything is wrong.

rgommers commented on benchmarks/benchmarks/stats.py:
```diff
@@ -228,3 +228,40 @@ def setup(self, dist_name, loc_fixed, scale_fixed, shape1_fixed,
+        self.a = a
+
+    def time_ks(self, alternative,mode, size):
+        statistic, pvalue = stats.kstest(self.a,'norm', alternative=alternative,mode=mode,N=size)
```
minor: missing a few spaces after commas, plus line is a bit long. can you fix that and put this on two lines or remove `statistic, pvalue =` (don't need to assign values here)?

rgommers commented on benchmarks/benchmarks/stats.py:
```diff
@@ -228,3 +228,40 @@ def setup(self, dist_name, loc_fixed, scale_fixed, shape1_fixed,
+    params = [
+        ['two-sided', 'less', 'greater'],
+        ['exact', 'approx', 'asymp'],
+        [19,20,21]
```
minor: PEP8 would like spaces

rgommers commented on benchmarks/benchmarks/stats.py:
```diff
@@ -228,3 +228,40 @@ def setup(self, dist_name, loc_fixed, scale_fixed, shape1_fixed,
+    params = [
+        ['two-sided', 'less', 'greater'],
+        ['exact', 'asymp'],
+        [(20,21),(20,20),(21,20)]
```
minor: PEP8 would like spaces

rgommers replied:
other question: is it necessary to test both `(20, 21)` and `(21, 20)`? I'd expect behaviour and performance to be symmetrical in the two input arrays

Balthasar-eu replied:
Yes, they are symmetrical. I will remove one version.

rgommers commented on benchmarks/benchmarks/stats.py:
```diff
@@ -228,3 +228,40 @@ def setup(self, dist_name, loc_fixed, scale_fixed, shape1_fixed,
+        self.b = b
+
+    def time_ks2(self, alternative,mode, size):
+        statistic, pvalue = stats.ks_2samp(self.a,self.b, alternative=alternative,mode=mode)
```
same comment as for the other benchmark above

rgommers commented on scipy/stats/stats.py:
```diff
@@ -6636,32 +6636,29 @@ def _count_paths_outside_method(m, n, g, h):
+            term = B[j] * bin
+            num_paths += term
+        return num_paths
+    except (RuntimeWarning,OverflowError):
```
catching a warning and turning it into an error is a bit odd. A `RuntimeWarning` could be caused by a lot of things, so this doesn't look quite right. What is the exact cause of the warning you're trying to catch, and can you deal with that another way?

rgommers replied:
If it's the `if not np.isfinite(term):`, maybe a more granular method would be using `with np.errstate(invalid='raise'):`?

Balthasar-eu replied:
Yeah, I thought that might be problematic, but I did not think of that solution. I try to catch the runtime warning that comes from the overflow. The outer function catches a FloatingPointError which was raised before by checking np.isfinite. The whole error handling could probably improved by using np.errstate in the outer function as well.

pvanmulbregt replied:
Should it be `np.errstate(invalid='raise')` or `np.errstate(over='raise')`? (The FloatingPointError was only raised explicitly if the computation overflowed.)
Does this catch overflows which occur in `scipy.special.binom` or only Python code under the `np.errstate` context?

Balthasar-eu replied:
It does not seem to catch overflows in special.binom. `np.errstate(invalid='raise')` produces errors instead of warnings, as can be seen in the test cases, so I think it should be correct.

pvanmulbregt replied:
The computation involving binomials doesn't have to deal with invalid operations, only overflows arising from binomial coefficients being larger than `np.info(float).max` (or multiplication of such binomials by `B[j]`).  E.g. `scipy.special.binom(1100, 512)` returns `np.inf`.  If the j-th term is infinite, the computation bails out, returning FloatingPointError(). OverflowError() may have also been a possible Exception to use for this situation.
So the context should be:
  `with np.errstate(over='raise'):`

rgommers commented on scipy/stats/stats.py:
```diff
@@ -6636,32 +6636,29 @@ def _count_paths_outside_method(m, n, g, h):
+        # Compute the number of path extensions...
+        num_paths = 0
+        for j in range(lxj):
+            bin = special.comb((m-xj[j]) + (n - j), n-j,exact=True)
```
PEP8: another space missing

rgommers reviewed (commented):
Thanks @Balthasar-eu. Overall looks good, some minor and one important comment. 

Also, how performance-critical is this for you? The double for-loop looks like a good candidate for moving to Cython or Pythran.

Balthasar-eu commented:
Thanks for your comments. The kstest is currently my performance bottleneck, these changes make it fast enough for me, but I have multiple million pairs to test, so every speedup directly cuts down the execution time.
I also tried to exchange the double loop and the last single loop with numpy array operations, but `special.comb` does not accept arrays as input when using `exact=True` and the double loop is triangular, not quadratic, so I gave up on that. I don't have experience with Cython or Pythran, but I can take a look into it next year.

Balthasar-eu commented:
I did some more testing and it seems that using `special.comb(exact=True)` introduces a performance regression for large input datasets. It seems to happen when the result of the function is larger than the int maximum. However the main speed reduction actually comes from the `np.round` functions around the `special.binom` calls. I think are unnecessary, and were probably originally added just as precautions for floating point inaccuracies, but as even `special.binom(1000,500) == np.round(special.binom(1000,500))`, which is very close to the max for floats returns True, I don't think the output of `special.binom` needs to be rounded.

So for the best performance I can revert to use binom instead of comb and just remove the round functions around the binom calls. I could also additionally make a check if the output of special.binom is likely below the int max and if so use special.comb, which is still ~5x faster for smaller numbers, but that might make the code a bit too messy.

rgommers commented:
>  I think are unnecessary, and were probably originally added just as precautions for floating point inaccuracies, but as even `special.binom(1000,500) == np.round(special.binom(1000,500))`, which is very close to the max for floats returns True, I don't think the output of `special.binom` needs to be rounded.

`(1000, 500)` is too large to check this, for results over ~1e16 the rounding doesn't do anything anymore. Here is the kind of thing it was meant for:
```
>>> res = special.binom(121, 11)
>>> res
1276749965026536.2
>>> (res - np.round(res)) / res
1.958096783615764e-16
```
The relative correction from rounding is always going to be close to a few ULP it looks like though, so it looks like rounding isn't all that useful.

This will take a while to review. One additional issue I noticed is that `comb` and `binom` docstrings are really bad, they don't even explain why we have both and what the performance/accuracy trade-offs are.

Balthasar-eu commented:
I made some actual comparisions for error and runtime between `special.binom`, `special.comb` and `special.comb(exact=True)`. As expected, the runtime for `special.comb(exact=True)` makes a jump when the result increases over the int max (denoted by the vertical line). The relative error (calculated by `binom(i,i//2) - comb(i,i//2,exact=True)/binom(i,i//2)` ) stays small throughout the whole range, but the absolute error gets very large. Rounding only has a positive effect in the region, where the error is smaller than 1, but also only if it is less than 0.5, otherwise the error actually gets increased. The last image shows all possible binom values until 1000, not just cases where k is half of n. In this plot are 210 values for which the error is not null, but smaller than 1, for 22 of these cases the error is greater than 0.5, so rounding makes the error worse. Finally there are 117520 cases where the error is larger than 1. So I think it is fine to remove the rounding statements, since for most of the result space they have no beneficial effect, and for the cases, where a beneficial effect exists, the effect is small.

![binom_comparison](https://user-images.githubusercontent.com/36669346/103418245-226dea00-4b8e-11eb-99a3-a51e2a137c06.png)

rgommers commented:
Thanks for the detailed investigation @Balthasar-eu. That's a convincing argument for it being fine to remove `round`.

Balthasar-eu commented:
Great, here is the cost of rounding just for completeness sake, because I forgot to add it in the first plot. The average for non rounded binom is 1.3µs and rounded 5.12µs on my machine:
![round_cost](https://user-images.githubusercontent.com/36669346/103987542-ed185c00-518c-11eb-9c85-8b7701e92e56.png)

rgommers commented:
@pvanmulbregt you did quite a bit of work on the KS-test functions before, would you be interested in reviewing this PR?

pvanmulbregt commented on scipy/stats/stats.py:
```diff
@@ -6637,31 +6637,23 @@ def _count_paths_outside_method(m, n, g, h):
-        Bj = np.round(special.binom(xj[j] + j, j))
-        if not np.isfinite(Bj):
-            raise FloatingPointError()
+        Bj = special.binom(xj[j] + j, j)
```
If FloatingPointError is not explicitly raised, the the docstring for `_count_paths_outside_method` needs to be updated to reflect that.

Balthasar-eu replied:
Okay, should the section just be deleted, or should be mentioned that the function is wrapped with `np.errstate`?

pvanmulbregt replied:
If `_count_paths_outside_method` doesn't raise FloatingPointError, just remove it from the docstring.  How other functions call it or wrap it doesn't need to get mentioned in its docstring.

Balthasar-eu commented:
Sorry for the long break, I switched projects and forgot a bit about this pull request.  I rebased and removed the superflouous docstring.

I saw there is a pull request (https://github.com/scipy/scipy/pull/14314) now to implement _count_paths_outside_method by reimplementing it in pythran, but the improvements from this pull request are not included and would still benefitial (remove rounding and set failstate to error so all the checks are unnecessary)

Balthasar-eu commented:
I think there might have been an issue with rebasing, but I am not sure how to fix it.

ev-br commented:
When rebasing, one recommendation  is to locate your first commit, let's call its hash HASH. Then rebase via `$ git rebase -I HASH^` --- notice the carpet, which means "one commit before HASH".

mdhaber commented:
I merged main here, being careful to include the call to the new-ish `_compute_outer_prob_inside_method`. It was hairy, so I am noting some details here:

- There were no merge conflicts with the test or benchmarks
- `stats.py` is now `_stats_py.py`, so I had to figure out the merge conflicts myself. Since this PR only changes `_count_paths_outside_method` and `_attempt_exact_2kssamp`, I took a look at the differences in those functions between the commit this PR is _based on_ and the latest changes in main. There were two differences: the call to `_count_paths_outside_method` and a new `with np.errstate(over='raise'):` around the call to `_count_paths_outside_method`.
- I ignored changes to `stats.py`.
- I replaced the `_count_paths_outside_method` and `_attempt_exact_2kssamp` in `_stats_py.py` with those from this PR.
- I added back the changes to those methods noted in step 2.

There is one test failing locally related to the resolution of gh-14019 by gh-16525. I'll have to investigate more closely to determine what to do.

@Balthasar-eu if you're still interested in continuing this, let me know. All you should need to do is fetch and merge the latest commit before continuing work. (Note: I saw some discussion of _rebasing_ above. If you'd like me to help review this, please do not rebase as we go forward. If needed, we can _merge_ with main. Different reviewers have different preferences, but my preference is to merge because rebasing erases the history, so I can't be sure of what has changed since I last looked. We can discuss rebasing - or more likely, just squash-merging - to clean up the history once review is essentially done.)

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7734,29 +7714,29 @@ def _attempt_exact_2kssamp(n1, n2, g, d, alternative):
+                    with np.errstate(over='raise'):
+                        num_paths = _count_paths_outside_method(n1, n2, g, h)
+                    bin = special.binom(n1 + n2, n1)
+                    if num_paths > bin:
```
```suggestion
                    if num_paths > bin or np.isinf(bin):
```
In the failing test, `num_paths` is finite but `bin` is not, and there is no overflow warning. 
``python3
from scipy import special
special.binom(369 + 881, 369)  # inf, no warning
```
In the (very unlikely) case that `num_paths` is only slightly smaller than the true value of `binom(n1 + n2, n1)`, this will incorrectly report that the p-value is zero. I think it would be safer to make `saw_fp_error = True` and revert to the normal approximation as the code would do now. 

Alternatively, we could consider computing with logs, e.g. `logbinom(n, k) = loggamma(n+1) - loggamma(n-k+1) - loggamma(k+1)`, although that should be a separate enhancement.

mdhaber replied:
@Balthasar-eu I went ahead and pushed this change, but do you agree?

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7633,7 +7633,6 @@ def _compute_prob_outside_square(n, h):
 
 def _count_paths_outside_method(m, n, g, h):
     """Count the number of paths that pass outside the specified diagonal.
```
Let's not make unnecessary whitespace changes.
```suggestion
    """Count the number of paths that pass outside the specified diagonal.

```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7644,31 +7643,22 @@ def _count_paths_outside_method(m, n, g, h):
         g is greatest common divisor of m and n
     h : integer
         0 <= h <= lcm(m,n)
```
```suggestion
        0 <= h <= lcm(m,n)

```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7644,31 +7643,22 @@ def _count_paths_outside_method(m, n, g, h):
     point (x, y) along the path, satisfy:
       m*y <= n*x - h*g
     The paths make steps of size +1 in either positive x or positive y
     directions.
```
```suggestion
    directions.

```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7644,31 +7643,22 @@ def _count_paths_outside_method(m, n, g, h):
     We generally follow Hodges' treatment of Drion/Gnedenko/Korolyuk.
     Hodges, J.L. Jr.,
     "The Significance Probability of the Smirnov Two-Sample Test,"
     Arkiv fiur Matematik, 3, No. 43 (1958), 469-86.
```
```suggestion
    Arkiv fiur Matematik, 3, No. 43 (1958), 469-86.

```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7691,40 +7681,30 @@ def _count_paths_outside_method(m, n, g, h):
 
 
 def _attempt_exact_2kssamp(n1, n2, g, d, alternative):
     """Attempts to compute the exact 2sample probability.
```
```suggestion
    """Attempts to compute the exact 2sample probability.

```

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7691,40 +7681,30 @@ def _count_paths_outside_method(m, n, g, h):
-
     n1, n2 are the sample sizes
     g is the gcd(n1, n2)
     d is the computed max difference in ECDFs
```
```suggestion
    d is the computed max difference in ECDFs

```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -4117,8 +4117,9 @@ def test_some_code_paths(self):
-        assert_raises(FloatingPointError, _count_paths_outside_method, 1100, 1099, 1, 1)
-        assert_raises(FloatingPointError, _count_paths_outside_method, 2000, 1000, 1, 1)
+        with np.errstate(invalid='raise'):
+            assert_raises(FloatingPointError, _count_paths_outside_method, 1100, 1099, 1, 1)
```
PEP8
```suggestion
            assert_raises(FloatingPointError, _count_paths_outside_method,
                          1100, 1099, 1, 1)
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -4117,8 +4117,9 @@ def test_some_code_paths(self):
-        assert_raises(FloatingPointError, _count_paths_outside_method, 2000, 1000, 1, 1)
+        with np.errstate(invalid='raise'):
+            assert_raises(FloatingPointError, _count_paths_outside_method, 1100, 1099, 1, 1)
+            assert_raises(FloatingPointError, _count_paths_outside_method, 2000, 1000, 1, 1)
```
```suggestion
            assert_raises(FloatingPointError, _count_paths_outside_method, 
                          2000, 1000, 1, 1)
```

mdhaber commented on benchmarks/benchmarks/stats.py:
```diff
@@ -664,3 +664,41 @@ def setup(self, n_size):
+        self.a = a
+
+    def time_ks(self, alternative, mode, size):
+        stats.kstest(self.a, 'norm', alternative=alternative, mode=mode, N=size)
```
PEP8
```suggestion
        stats.kstest(self.a, 'norm', alternative=alternative, 
                     mode=mode, N=size)
```

mdhaber reviewed (commented):
Some PEP8 / style fixes.

mdhaber commented on scipy/stats/_stats_py.py:
```diff
@@ -7650,12 +7650,6 @@ def _count_paths_outside_method(m, n, g, h):
     p : float
         The number of paths that go low.
         The calculation may overflow - check for a finite answer.
```
```suggestion
        The calculation may overflow - check for a finite answer.

```

mdhaber commented on benchmarks/benchmarks/stats.py:
```diff
@@ -664,3 +664,42 @@ def setup(self, n_size):
+        self.a = a
+
+    def time_ks(self, alternative, mode, size):
+        stats.kstest(self.a, 'norm', alternative=alternative, 
```
```suggestion
        stats.kstest(self.a, 'norm', alternative=alternative,
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -4117,8 +4117,11 @@ def test_some_code_paths(self):
+        with np.errstate(invalid='raise'):
+            assert_raises(FloatingPointError, _count_paths_outside_method,
+                          1100, 1099, 1, 1)
+            assert_raises(FloatingPointError, _count_paths_outside_method, 
```
```suggestion
            assert_raises(FloatingPointError, _count_paths_outside_method,
```

mdhaber commented on scipy/stats/tests/test_stats.py:
```diff
@@ -4117,8 +4117,11 @@ def test_some_code_paths(self):
 
-        assert_raises(FloatingPointError, _count_paths_outside_method, 1100, 1099, 1, 1)
-        assert_raises(FloatingPointError, _count_paths_outside_method, 2000, 1000, 1, 1)
+        with np.errstate(invalid='raise'):
+            assert_raises(FloatingPointError, _count_paths_outside_method,
+                          1100, 1099, 1, 1)
+            assert_raises(FloatingPointError, _count_paths_outside_method,
+                          2000, 1000, 1, 1)
```
If you want to make sure that the `invalid` `RuntimeWarning` warning is raised.
```suggestion
        message = "invalid value encountered"
        with pytest.warns(RuntimeWarning, match=message):
            _count_paths_outside_method(1100, 1099, 1, 1)
        message = "invalid value encountered"
        with pytest.warns(RuntimeWarning, match=message):
            _count_paths_outside_method(2000, 1000, 1, 1)
```
Or, you could make sure that this leads to the warning `ks_2samp: Exact calculation unsuccessful. Switching to method=asymp.` when we attempt to use `ks_2samp` on samples of these sizes.

mdhaber reviewed (approved):
This PR removes `np.round`s and replaces finiteness checks with `np.errstate(invalid="raise", over="raise")` from the `_count_paths_outside_method` loops, improving performance considerably. I didn't review it myself, but it sounds like the case for not needing `np.round` [was strong](https://github.com/scipy/scipy/pull/13265#issuecomment-754252601). If the benchmark runs well and CI comes back green, I think this is ready, although there is one optional comment to consider.

@pvanmulbregt we've missed you! care to take one last look at this one?

mdhaber commented:
Thanks @Balthasar-eu!
